src/session/session_dhandle.c

```diff
@@ -404,13 +404,33 @@ static int
 __session_find_shared_dhandle(WT_SESSION_IMPL *session, const char *uri, const char *checkpoint)
 {
     WT_DECL_RET;
+    bool is_checkpoint;
 
-    WT_WITH_HANDLE_LIST_READ_LOCK(session,
-      if ((ret = __wt_conn_dhandle_find(session, uri, checkpoint)) == 0)
-        WT_DHANDLE_ACQUIRE(session->dhandle));
-
-    if (ret != WT_NOTFOUND)
-        return (ret);
+    /*
+     * If we get called as part of the checkpoint, and the handle is the checkpoint to be created,
+     * it is guaranteed that the handle won't exist in the shared dhandle list. Skip searching
+     * through the list, and directly insert into it.
+     */
+    is_checkpoint = (WT_SESSION_IS_CHECKPOINT(session) && checkpoint != NULL &&
+      WT_PREFIX_MATCH(checkpoint, WT_CHECKPOINT));
+    if (!is_checkpoint) {
+        WT_WITH_HANDLE_LIST_READ_LOCK(session,
+          if ((ret = __wt_conn_dhandle_find(session, uri, checkpoint)) == 0)
+            WT_DHANDLE_ACQUIRE(session->dhandle));
+
+        if (ret != WT_NOTFOUND)
+            return (ret);
+    } else {
+#ifdef HAVE_DIAGNOSTIC
+        /*
+         * Let's check our assumption that this handle is not already there in the shared list is
+         * actually correct.
+         */
+        WT_WITH_HANDLE_LIST_READ_LOCK(session,
+          if ((ret = __wt_conn_dhandle_find(session, uri, checkpoint)) == 0)
+            WT_ASSERT(session, ret != WT_NOTFOUND));
```

Comment:
I don't think this is right. The assertion is not meaningful because as written `ret` is guaranteed to be equal to 0 and we want to assert that `ret == WT_NOTFOUND`. I think the `if` should not be part of the statement in the `WT_WITH_HANDLE_LIST_READ_LOCK`. It should look like:
`WT_WITH_HANDLE_LIST_READ_LOCK(session, ret = __wt_conn_dhandle_find(session, uri, checkpoint)))`
`WT_ASSERT(session, ret == WT_NOTFOUND);`